### PR TITLE
gh-74929: locals() documentation update for PEP 667

### PR DESCRIPTION
**Pull request by ncoghlan**
Proposed documentation wording taken from the
withdrawn PEP 558 (which had switched to the
Python level semantics proposed in PEP 667 prior
to its withdrawal).

----
📚 Documentation preview 📚: https://cpython-previews--118265.org.readthedocs.build/

* Issue: gh-74929

### CODE REVIEW

**Comment by ncoghlan** [comment]
DO-NOT-MERGE label added, as this should only be merged after the PEP 667 implementation in https://github.com/python/cpython/pull/115153 has been completed and merged.

File: Doc/library/functions.rst [hunk 1]
```diff
@@ -1041,14 +1040,35 @@ are always available.  They are listed here in alphabetical order.
...
+    object.
+
+    At function scope (including for generators and coroutines), each call to
+    ``locals()`` instead returns a fresh dictionary containing the current
+    bindings of the function's local variables and any nonlocal cell references.
+    In this case, name binding changes made via the returned dict are *not*
+    written back to the corresponding local variables or nonlocal cell
+    references, and binding, rebinding, or deleting local variables and nonlocal
+    cell references does *not* affect the contents of previously returned
+    dictionaries.
```
Comment (ncoghlan) [inline Doc/library/functions.rst:1070]: I need to add a `:versionchanged:` note here.

**Comment by nineteendo** [comment]
Wouldn't it better to convert the pull request back to a draft? GitHub automatically prevents you from merging those.

**Comment by ncoghlan** [comment]
> Wouldn't it better to convert the pull request back to a draft? GitHub automatically prevents you from merging those.

Good point, done (enough years of experience where draft PRs weren't available that I still forget to use them when appropriate)

File: Doc/library/functions.rst (hunk 1 above)
Comment (ncoghlan) [reply Doc/library/functions.rst:1070]: This has been added.

File: Doc/library/functions.rst [hunk 2]
```diff
@@ -1041,14 +1040,44 @@ are always available.  They are listed here in alphabetical order.
...
+
+   .. versionchanged:: 3.13
+      In previous versions, the semantics of mutating the mapping object
+      returned from this function were formally undefined. In CPython
+      specifically, the mapping returned at function scope could be
+      implicitly refreshed by other operations, such as calling ``locals()``
+      again. Obtaining the legacy CPython behaviour now requires explicit
+      calls to update the initially returned dictionary with the results
+      of subsequent calls to ``locals()``.
```
Comment (ncoghlan) [inline Doc/library/functions.rst:1079]: I plan to eventually replace the latter two sentences here with something like "For details, refer to the 'Python 3.13 porting guide'" (with a link to the relevant part of the 3.13 What's New document). That will be in a later PR that adds an initial version of the What's New update for PEP 667, though.

**Pull request merged**